wallet: Move restorewallet() logic to the wallet section

w0xlt opened the pull request:
Currently `restorewallet()` logic is written in the RPC layer and it can´t be reused by GUI. So it moves this to the wallet section and then, GUI can access it.

This is necessary to implement the "Restore Wallet" menu item in the GUI (which is already implemented  in https://github.com/bitcoin-core/gui/pull/471 ). 

This commit also simplifies error handling and adds a new behavior: if the restore fails, the invalid wallet folder is removed.

shaavan reviewed (commented):
Concept ACK

Wow, this PR was a tough nut to crack! It took me quite a while to understand it thoroughly. That’s why I have attached my notes so that other reviewers might have an easier time reviewing this PR.

I agree with the changes introduced in this PR. Though I have some suggestions that might help improve this PR.
1. I suggest you split the PR into two commits: First, introduce the **RestoreWallet** function, and second, use it.
2. Due to the difference introduced in implementation, [this](https://github.com/bitcoin/bitcoin/pull/23721/files#diff-7b13e03c457b62251e26d2c366180c262a3fe9da45995277a9a9c1c5abbaae6cR218-R222) snippet of code is getting duplicated. Though I can’t find a solution to avoid this, it also doesn’t feel intuitively correct. Maybe other reviewers might have some suggestions to improve upon it.

**Notes:**

- Introduced new pure virtual function `restoreWallet` in `src/interfaces/wallet.h` file. This function is then overridden in `src/wallet/interfaces.h` file.
- This new function allows the creation of a new wallet using backed-up wallet’s data. For this purpose, it uses the (newly introduced) function **RestoreWallet.**
- **RestoreWallet** is defined under `src/wallet/wallet.h` file. This function does preliminary checks for:
    1. If `backup_file` exist.
    2. If first, `wallet_path` (i.e., where wallet should be created) is empty, and, second, the directory can be created there.
After these steps, the **LoadWallet** function is called, creating a wallet variable. Finally, if the **wallet** variable does not exist (`if(!wallet)`), the `wallet_file` and `wallet_path` are removed from the system. Finally, the **wallet** variable is returned if everything works out correctly.
- The **LoadWalletHelper** in `src/wallet/rpc/util.cpp`, which used to return {wallet, warning} has now been reduced to **HandleWalletError**. Reason:
    1. This function had two use cases: for `restoreWallet()` in `src/wallet/rpc/backup.cpp` file, and for `loadWallet()` in `src/wallet/rpc/wallet.cpp` file.
    2. Now, the `restoreWallet()` function was updated to use a more suitable **RestoreWallet** function, and `loadWallet()` is still using the **LoadWallet** function.
    3. This change leads to duplicating the following steps for both functions and removing them from the **LoadWalletHelper** function.
```
    DatabaseOptions options;
    DatabaseStatus status;
    options.require_existing = true;
    bilingual_str error;
    std::vector<bilingual_str> warnings;
```
- Finally, this PR introduced some minor changes to the `test/functional/wallet_backup.py` file:
    1. Changed the assert_raises_rpc_error code from -8 (**RPC_INVALID_PARAMETER**) to -4 (**RPC_WALLET_ERROR**)
    2. Updated code to make the error message more verbose, better explaining the cause of the error.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -357,6 +357,35 @@ std::shared_ptr<CWallet> CreateWallet(WalletContext& context, const std::string&
     return wallet;
 }
 
+std::shared_ptr<CWallet> RestoreWallet(WalletContext& context, const std::string& backup_file, const std::string& wallet_name, std::optional<bool> load_on_start, const DatabaseOptions& options, DatabaseStatus& status, bilingual_str& error, std::vector<bilingual_str>& warnings)
```
In 33763d02ef1011ae71143c5e6e313c72d9b790a6 "refactor: Move restorewallet() logic to the wallet section"

`options` does not need to be an argument to this function. The `DatabaseOptions` for loading a restored wallet are always going to be the same so you can make that object inside of this function.

w0xlt replied:
Done in 4807f73

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -357,6 +357,35 @@ std::shared_ptr<CWallet> CreateWallet(WalletContext& context, const std::string&
+    auto wallet = LoadWallet(context, wallet_name, load_on_start, options, status, error, warnings);
+
+    if (!wallet) {
+        fs::remove(wallet_file);
```
In 33763d02ef1011ae71143c5e6e313c72d9b790a6 "refactor: Move restorewallet() logic to the wallet section"

This is a change in behavior. I would prefer to have this added in its own commit.

w0xlt replied:
Done in 2e5195e
Also added a more one test for this new behavior.

achow101 commented on src/wallet/rpc/backup.cpp:
```diff
@@ -1879,27 +1879,19 @@ RPCHelpMan restorewallet()
+    const std::shared_ptr<CWallet> wallet = RestoreWallet(context, fs::PathToString(backup_file), wallet_name, load_on_start, options, status, error, warnings);
 
-    auto [wallet, warnings] = LoadWalletHelper(context, request.params[2], wallet_name);
+    HandleWalletError(wallet, status, error);
```
In 33763d02ef1011ae71143c5e6e313c72d9b790a6 "refactor: Move restorewallet() logic to the wallet section"

`HandleWalletError` does not handle `FAILED_BAD_PATH` or `FAILED_ALREADY_EXISTS`, which results in a slightly incorrect `RPC_WALLET_ERROR` code. There should be additional error handling for those cases to return better error codes.

w0xlt replied:
Done in a9f9a83, which introduced additional error handling.

achow101 commented on src/wallet/rpc/util.cpp:
```diff
@@ -122,16 +122,8 @@ std::string LabelFromValue(const UniValue& value)
 }
 
-std::tuple<std::shared_ptr<CWallet>, std::vector<bilingual_str>> LoadWalletHelper(WalletContext& context, UniValue load_on_start_param, const std::string wallet_name)
+void HandleWalletError(const std::shared_ptr<CWallet> wallet, DatabaseStatus& status, bilingual_str& error)
```
In 33763d02ef1011ae71143c5e6e313c72d9b790a6 "refactor: Move restorewallet() logic to the wallet section"

It would be easier to review this diff if this refactor were in its own commit.

w0xlt replied:
Done in a9f9a83.

w0xlt committed:
refactor: Implement restorewallet() logic in the wallet section

Currently restorewallet() logic is written in the RPC layer
and it can´t be reused by GUI. So it reimplements this in the
wallet and interface sections and then, GUI can access it.

w0xlt committed:
refactor: Move restorewallet() RPC logic to the wallet section

It also simplifies restorewallet() and loadwallet() RPC error handling.

Rspigler commented:
As a followup, documentation should be updated

w0xlt committed:
refactor: remove the wallet folder if the restore fails

achow101 commented on src/wallet/db.h:
```diff
@@ -220,6 +220,7 @@ enum class DatabaseStatus {
     FAILED_LOAD,
     FAILED_VERIFY,
     FAILED_ENCRYPT,
+    FAILED_INVALID_BACKCUP_FILE,
```
In 4053a210a6126f85779e5ccab248d13a74d27897 "refactor: Move restorewallet() RPC logic to the wallet section"

nit: typo

```suggestion
   FAILED_INVALID_BACKUP_FILE,
```

w0xlt replied:
Fixed in abbb7ec

achow101 commented:
ACK 62fa61fa4a33ff4d108a65d656ffe2cac4330824

shaavan reviewed (approved):
crACK 62fa61fa4a33ff4d108a65d656ffe2cac4330824

Great work, @w0xlt!

MarcoFalke commented:
None of this is a refactor. Every commit is a behavior change.
